### PR TITLE
Move generation and application of patches into a different package

### PR DESCRIPTION
**Pull request by algomaster99**
Reference: #42 

The changes create a new package `com.diffmin.patch` which contains the logic for generation and application of patches.

Next goal can be to move the `App` class to the `util` package as it only contains utility functions now.

### CODE REVIEW

File: src/main/java/com/diffmin/patch/Generation.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+                updatePatches.add(update(operation.getSrcNode(), operation.getDstNode()));
+            } else if (operation.getAction() instanceof Insert) {
+                insertPatches.add(insert(operation.getSrcNode(), mapping));
+                insert(operation.getSrcNode(), mapping);
+            }
+        }
+    }
+
+    private static CtElement delete(CtElement removedNode) {
+        return removedNode;
```
Comment (algomaster99) [inline src/main/java/com/diffmin/patch/Generation.java:46]: I could have directly added the node to `deletePatches` here but I think writing it this way will help us in writing logic for `MoveOperation`.

Moreover, creation of this method is just for syntactical sugar. I wanted to have separate methods for generating the patches corresponding to the `Operation`s. An implementation like this seems clearer to me.

File: src/main/java/com/diffmin/patch/Generation.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    private static ImmutableTriple<Integer, CtElement, CtElement> insert(
+            CtElement insertedNode, SpoonMapping mapping) {
+
+        InsertionUtil iUtil =
+                new InsertionUtil() {
+
+                    /** Computes the index at which the `insertedNode` has to be inserted. */
+                    @Override
+                    public int getSrcNodeIndex(CtElement srcNode) {
```
Comment (algomaster99) [inline src/main/java/com/diffmin/patch/Generation.java:61]: This function is a part of the insert patch specifically (not update and delete) so I wanted to keep this method inside `insert` logic only. So creating an anonymous class made sense to as it isn't accessible by other methods for generating patches.

File: src/main/java/com/diffmin/patch/Application.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.commons.lang3.tuple.ImmutableTriple;
+import spoon.reflect.code.CtExpression;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.code.CtStatement;
+import spoon.reflect.code.CtStatementList;
+import spoon.reflect.declaration.*;
+
+/** Class for applying patches. */
+public class Application {
+    private Set<CtElement> deletePatches;
```
Comment (algomaster99) [inline src/main/java/com/diffmin/patch/Application.java:14]: Note that I converted this into a `HashSet` just to avoid any duplicated.

File: src/main/java/com/diffmin/patch/Generation.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.stream.IntStream;
+import org.apache.commons.lang3.tuple.ImmutableTriple;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.code.CtStatementList;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtExecutable;
+
+/** Class for generating patches. */
+public class Generation {
```
Comment (slarse) [inline src/main/java/com/diffmin/patch/Generation.java:22]: Same comment as for `Application`, I think this class would be better named something like `PatchGenerator`.

File: src/main/java/com/diffmin/Main.java [hunk 5]
```diff
@@ -16,12 +21,23 @@
...
-    public static CtModel patchAndGenerateModel(File prevFile, File newFile)
-            throws FileNotFoundException {
-        App app = new App();
+    static CtModel patchAndGenerateModel(File prevFile, File newFile) throws FileNotFoundException {
         Pair<Diff, CtModel> diffAndModel = App.computeDiff(prevFile, newFile);
-        app.generatePatch(diffAndModel.getFirst());
-        app.applyPatch();
+
+        // Generate patches
+        Generation patchGenerationFactory = new Generation();
```
Comment (slarse) [inline src/main/java/com/diffmin/Main.java:28]: This is not a factory in the sense that the word is most often used, see [the factory pattern](https://en.wikipedia.org/wiki/Factory_method_pattern). That could be confusing.

File: src/main/java/com/diffmin/patch/Generation.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    public final Set<CtElement> deletePatches = new HashSet<>();
+    public final Set<Pair<CtElement, CtElement>> updatePatches = new HashSet<>();
+    public final Set<ImmutableTriple<Integer, CtElement, CtElement>> insertPatches =
+            new HashSet<>();
```
Comment (slarse) [inline src/main/java/com/diffmin/patch/Generation.java:26]: Two problems here:

### `HashSet` for order-dependent collections

Using a `HashSet` for insert patches is probably not a good idea. If I'm not mistaken, insert patches are dependent on the order of operations, as the computed index in a given list is only correct if all preceding insertions have already been performed. Thus, if you really want a set here, you should use one that preserves insertion order (a property that `HashSet` does not have).

However, you shouldn't use sets here. For `deletePatches`, you can have incorrect behavior as two equal nodes (say `int a = 2` appearing in two different places) would interfere with each other (only one can be added to the set). Thus, you might lose deletions.

For `updatePatches`, your `Pair` class doesn't implement `equals` and `hashCode`, and so its set properties in a `HashSet` will simply be those of `Object` (compare by reference). Ironically, this is the only one that will actually work as intended!

Finally, for `insertPatches`, you have the same problem as for `deletePatches`, assuming that `ImmutableTriple` has a reasonable `equals` and `hashCode` implementation.

To summarize: I would just use lists here.

### Public mutable fields

Public, mutable fields is a big no-no. When you mark a collection as final, that only applies to the reference in the field, and not what it points to. For example, I can't do this:

```java
Generation generation = new Generation();
generation.deletePatches = new HashSet<>(); // error, deletePatches is final
```

but I can do this:

```java
Generation generation = new Generation();
generation.deletePatches.add(someCtElement); // this is fine
```

Not only does it allow me to see the internals of the class from the outside, it allows me to modify them!

**Review by slarse — CHANGES_REQUESTED** [review]
Overall a very good refactoring. One thing that you absolutely shouldn't do in a refactoring, however, is to change behavior, which you've done with changing lists to sets in the patch generation. See my comments there.

File: src/main/java/com/diffmin/patch/Application.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.Set;
+import org.apache.commons.lang3.tuple.ImmutableTriple;
+import spoon.reflect.code.CtExpression;
+import spoon.reflect.code.CtInvocation;
+import spoon.reflect.code.CtStatement;
+import spoon.reflect.code.CtStatementList;
+import spoon.reflect.declaration.*;
+
+/** Class for applying patches. */
+public class Application {
```
[the remark this answers is not in the document]
Comment (algomaster99) [reply src/main/java/com/diffmin/patch/Application.java:13]: Renaming it to `Applier` as `patch` is the name of the package so `PatchApplier` just sounds repititive to me.

[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/com/diffmin/patch/Application.java:13]: I get that reasoning, but take into account that you don't use the qualified name when using the class. It just looks like this:

```
Applier applier = new Applier();
```

As opposed to:

```
PatchApplier applier = new PatchApplier();
```

I often add redundancy between package and class names for just that reason: the package name simply isn't often used explicitly other than for the import statement, and that doesn't help readability where the class is used.

File: src/main/java/com/diffmin/patch/Generation.java (hunk 6 above)
Comment (algomaster99) [reply src/main/java/com/diffmin/patch/Generation.java:26]: > a property that HashSet does not have

Oh, I thought `HashSet` preserves order while maintaing the uniqueness.

> For deletePatches, you can have incorrect behavior as two equal nodes (say int a = 2 appearing in two different places) would interfere with each other (only one can be added to the set). Thus, you might lose deletions.

I put some thought here and then I realised that the `CtElement` in the delete patch will have an attribute `parent` which will definitely differ so same statements can co-exist if the previously stated condition is met.

**Comment by algomaster99** [comment]
> however, is to change behavior, which you've done with changing lists to sets in the patch generation.

One feature / one PR. Hopefully, this gets engrained soon. Anyway, I changed this to a set to filter out the duplications in case they every arise. However, it requires much more thought than that so I shall revert those changes. I will make a separate PR for handling duplicate operations where I will address [this comment](https://github.com/SpoonLabs/diffmin/pull/56#discussion_r633255687) as well (the optimal way).

File: src/main/java/com/diffmin/patch/Generation.java (hunk 6 above)
Comment (slarse) [reply src/main/java/com/diffmin/patch/Generation.java:26]: > I put some thought here and then I realised that the CtElement in the delete patch will have an attribute parent which will definitely differ so same statements can co-exist if the previously stated condition is met.

This doesn't work for two reasons.

1. `parent` is not part of the equality check of a `CtElement`. The semantics of `CtElement.equals(CtElement)` is to check if they two _subtrees rooted in those elements_ are equal.
2. Even if `parent` was part of the equality check, what about this (`a++` nodes are identical):

```java
int a = 2;
a++;
a++;
```

Comment (algomaster99) [reply src/main/java/com/diffmin/patch/Generation.java:26]: Oh, okay. I didn't think through. I did not know if `equals` was overridden. I assumed it would have a default behaviour. 

> Even if parent was part of the equality check, what about this (a++ nodes are identical):

Great example! Thanks :)

Comment (slarse) [reply src/main/java/com/diffmin/patch/Generation.java:26]: > Oh, okay. I didn't think through. I did not know if equals was overridden. I assumed it would have a default behaviour.

Default behavior for equals is to compare by reference (see [Object.equals()](https://docs.oracle.com/javase/8/docs/api/java/lang/Object.html#equals-java.lang.Object-). So, for _any_ class that requires true equality comparison, one must override equals (which also entails overriding hashCode). Perhaps we should discuss this during the meeting, it's one of those fundamental Java things that one must be familiar with.

**Review by slarse — APPROVED** [review]
Nice, well done!

**Comment by algomaster99** [comment]
@slarse Next steps would be to shift the `App` class into the `util` package. Afterwards, I will write the `MoveOperation` logic. I have a prototype ready :)
